neo/Ledger/MemoryPool.cs

```diff
@@ -286,31 +287,42 @@ internal bool TryAdd(UInt256 hash, Transaction tx)
 
             if (_unsortedTransactions.ContainsKey(hash)) return false;
 
+            List<Transaction> removedTransactions = null;
             _txRwLock.EnterWriteLock();
             try
             {
                 _unsortedTransactions.Add(hash, poolItem);
 
                 SortedSet<PoolItem> pool = tx.IsLowPriority ? _sortedLowPrioTransactions : _sortedHighPrioTransactions;
                 pool.Add(poolItem);
-                RemoveOverCapacity();
+                RemoveOverCapacity(ref removedTransactions);
             }
             finally
             {
                 _txRwLock.ExitWriteLock();
             }
 
+            foreach (IMemoryPoolTxObserverPlugin plugin in Plugin.TxObserverPlugins)
+            {
+                plugin.AddedTransaction(poolItem.Tx);
+                if (removedTransactions != null)
+                    plugin.RemovedTransactions(MemoryPoolTxRemovalReason.CapacityExceeded, removedTransactions);
+            }
+
             return _unsortedTransactions.ContainsKey(hash);
         }
 
-        private void RemoveOverCapacity()
+        private void RemoveOverCapacity(ref List<Transaction> removedTransactions)
         {
             while (Count > Capacity)
             {
                 PoolItem minItem = GetLowestFeeTransaction(out var unsortedPool, out var sortedPool);
 
                 unsortedPool.Remove(minItem.Tx.Hash);
                 sortedPool.Remove(minItem);
+                if (removedTransactions == null)
```

Comment:
If is null, we should avoid this?

Reply:
It is null to delay performing the allocation. I realize that the c# collector for gen0 is very fast so this isn’t that more optimal, but since it will not be iterating the whole multiple times anyway right now, this is slightly more optimal.